src/target/sam3x.c

```diff
@@ -195,9 +228,152 @@ static size_t sam_flash_size(uint32_t cidr)
 	return 0;
 }
 
+static size_t sam_sram_size(uint32_t cidr)
+{
+	switch (cidr & CHIPID_CIDR_SRAMSIZ_MASK) {
+	case CHIPID_CIDR_SRAMSIZ_256K:
+		return 0x40000;
+	case CHIPID_CIDR_SRAMSIZ_384K:
+		return 0x60000;
+	default:
+		return 0;
+	}
+}
+
+struct samx7x_descr {
+	char product_code;
+	uint8_t product_id;
+	char pins;
+	uint32_t ram_size;
+	uint32_t flash_size;
+	uint8_t density;
+	char revision;
+};
+
+struct samx7x_descr samx7x_parse_id(uint32_t cidr, uint32_t exid) {
+
+	struct samx7x_descr descr = {0};
+
+	switch (cidr & CHIPID_CIDR_ARCH_MASK) {
+	case CHIPID_CIDR_ARCH_SAME70:
+		descr.product_code = 'E';
+		descr.product_id = 70;
+		break;
+	case CHIPID_CIDR_ARCH_SAMS70:
+		descr.product_code = 'S';
+		descr.product_id = 70;
+		break;
+	case CHIPID_CIDR_ARCH_SAMV71:
+		descr.product_code = 'V';
+		descr.product_id = 71;
+		break;
+	case CHIPID_CIDR_ARCH_SAMV70:
+		descr.product_code = 'V';
+		descr.product_id = 70;
+		break;
+	}
+
+	// A = Revision A, legacy version
+	// B = Revision B, current variant
+	switch (exid & CHIPID_CIDR_VERSION_MASK) {
+	case 0:
+		descr.revision = 'A';
+		break;
+	case 1:
+		descr.revision = 'B';
+		break;
+	default:
+		descr.revision = '_';
+		break;
+	}
+
+	// Q = 144 pins
+	// N = 100 pins
+	// J = 64 pins
+	switch (exid & CHIPID_EXID_SAMX7X_PINS_MASK) {
+	case CHIPID_EXID_SAMX7X_PINS_Q:
+		descr.pins = 'Q';
+		break;
+	case CHIPID_EXID_SAMX7X_PINS_N:
+		descr.pins = 'N';
+		break;
+	case CHIPID_EXID_SAMX7X_PINS_J:
+		descr.pins = 'J';
+		break;
+	}
+
+	descr.ram_size = sam_sram_size(cidr);
+	descr.flash_size = sam_flash_size(cidr);
+
+	// 21 = 2048 KB
+	// 20 = 1024 KB
+	// 19 = 512 KB
+	switch (descr.flash_size)
+	{
+	case 0x200000:
+		descr.density = 21;
+		break;
+	case 0x100000:
+		descr.density = 20;
+		break;
+	case 0x80000:
+		descr.density = 19;
+		break;
+	default:
+		descr.density = 0;
+		break;
+	}
+
+	return descr;
+}
+
+bool samx7x_probe(target *t)
+{
+	uint32_t cidr = target_mem_read32(t, SAM_CHIPID_CIDR);
+	uint32_t exid = 0;
+	if (cidr & CHIPID_CIDR_EXT) {
+		exid = target_mem_read32(t, SAM_CHIPID_EXID);
+	}
+
+	switch (cidr & CHIPID_CIDR_ARCH_MASK) {
+	case CHIPID_CIDR_ARCH_SAME70:
+	case CHIPID_CIDR_ARCH_SAMS70:
+	case CHIPID_CIDR_ARCH_SAMV71:
+	case CHIPID_CIDR_ARCH_SAMV70:
+		break;
+	default:
+		return false;
+	}
+
+	struct samx7x_descr descr = samx7x_parse_id(cidr, exid);
+	target_add_ram(t, 0x20000000, descr.ram_size);
```

Comment:
SRAM is at 0x20400000. You'll find dTCM at 0x20000000 only if configured in the GPNVM bits.